In [190]:
import pandas as pd
import numpy as np

### Importation des fichiers de données et transformation en dataframe

In [191]:
from pathlib import Path

# On part du dossier où se trouve le notebook, puis on remonte d'un niveau pour atteindre la racine du projet.
root_path = Path.cwd().parent 

# 2. Définir le chemin vers le dossier des données
data_folder = root_path / "data/raws"

print(f"🔍 Recherche des données dans : {data_folder}")

# 3. Créer un dictionnaire pour stocker nos DataFrames
# Clé = nom du fichier, Valeur = DataFrame chargé
all_dataframes = {}

# 4. Vérifier si le dossier existe
if data_folder.exists() and data_folder.is_dir():
    
    # On cherche tous les fichiers finissant par .csv dans le dossier data
    csv_files = list(data_folder.glob("*.csv"))
    
    if not csv_files:
        print("⚠️ Aucun fichier CSV trouvé dans le dossier 'data'.")
    else:
        print(f"✅ {len(csv_files)} fichiers trouvés. Chargement en cours...")
        
        for file_path in csv_files:
            try:
                # On utilise le nom du fichier (sans l'extension) comme clé
                df_name = file_path.stem 
                
                # Chargement du CSV
                df = pd.read_csv(file_path)
                
                # Stockage dans notre dictionnaire
                all_dataframes[df_name] = df
                
                print(f"   👉 {file_path.name} chargé ({len(df)} lignes)")
                
            except Exception as e:
                print(f"   ❌ Erreur sur {file_path.name} : {e}")
                
        print("\n✨ Chargement terminé !")
else:
    print(f"❌ Le dossier '{data_folder}' est introuvable. Vérifie ton arborescence.")


🔍 Recherche des données dans : c:\Users\johan\VS Code Wild 2\BRIEF_FAO_DataLab_Analyse_mondiale_de_la_sécurité_alimentaire\data\raws
✅ 5 fichiers trouvés. Chargement en cours...
   👉 fr_animaux.csv chargé (37166 lignes)
   👉 fr_céréales.csv chargé (891 lignes)
   👉 fr_population.csv chargé (175 lignes)
   👉 fr_sousalimentation.csv chargé (1020 lignes)
   👉 fr_vegetaux.csv chargé (104871 lignes)

✨ Chargement terminé !


### Analyse descriptive

In [192]:
print("👀 Inspection visuelle des premières lignes\n")
print("=" * 60)

for name, df in all_dataframes.items():
    print(f"📂 FICHIER : {name}")
    print(f"📏 Dimensions : {df.shape[0]} lignes, {df.shape[1]} colonnes")
    print("-" * 30)
    
    # Affichage des 5 premières lignes
    # On utilise display() pour une belle présentation dans le notebook
    display(df.head(10))
    
    print("\n" + "=" * 60 + "\n")

👀 Inspection visuelle des premières lignes

📂 FICHIER : fr_animaux
📏 Dimensions : 37166 lignes, 14 colonnes
------------------------------


,Code Domaine,Domaine,Code zone,Zone,Code Élément,Élément,Code Produit,Produit,Code année,Année,Unité,Valeur,Symbole,Description du Symbole
0,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2731,Viande de Bovins,2013,2013,Milliers de tonnes,134.00,S,Données standardisées
1,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5611,Importations - Quantité,2731,Viande de Bovins,2013,2013,Milliers de tonnes,6.00,S,Données standardisées
2,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5301,Disponibilité intérieure,2731,Viande de Bovins,2013,2013,Milliers de tonnes,140.00,S,Données standardisées
3,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5142,Nourriture,2731,Viande de Bovins,2013,2013,Milliers de tonnes,140.00,S,Données standardisées
4,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,645,Disponibilité alimentaire en quantité (kg/pers...,2731,Viande de Bovins,2013,2013,kg,4.59,Fc,Donnée calculée
5,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,664,Disponibilité alimentaire (Kcal/personne/jour),2731,Viande de Bovins,2013,2013,Kcal/personne/jour,27.00,Fc,Donnée calculée
6,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,674,Disponibilité de protéines en quantité (g/pers...,2731,Viande de Bovins,2013,2013,g/personne/jour,1.89,Fc,Donnée calculée
7,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,684,Disponibilité de matière grasse en quantité (g...,2731,Viande de Bovins,2013,2013,g/personne/jour,2.10,Fc,Donnée calculée
8,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2732,Viande d'Ovins/Caprins,2013,2013,Milliers de tonnes,150.00,S,Données standardisées
9,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5301,Disponibilité intérieure,2732,Viande d'Ovins/Caprins,2013,2013,Milliers de tonnes,150.00,S,Données standardisées




📂 FICHIER : fr_céréales
📏 Dimensions : 891 lignes, 14 colonnes
------------------------------


,Code Domaine,Domaine,Code zone,Zone,Code Élément,Élément,Code Produit,Produit,Code année,Année,Unité,Valeur,Symbole,Description du Symbole
0,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2511,Blé,2013,2013,Milliers de tonnes,5169,S,Données standardisées
1,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2805,Riz (Eq Blanchi),2013,2013,Milliers de tonnes,342,S,Données standardisées
2,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2513,Orge,2013,2013,Milliers de tonnes,514,S,Données standardisées
3,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2514,Maïs,2013,2013,Milliers de tonnes,312,S,Données standardisées
4,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2517,Millet,2013,2013,Milliers de tonnes,13,S,Données standardisées
5,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,202,Afrique du Sud,5511,Production,2511,Blé,2013,2013,Milliers de tonnes,1870,S,Données standardisées
6,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,202,Afrique du Sud,5511,Production,2805,Riz (Eq Blanchi),2013,2013,Milliers de tonnes,3,S,Données standardisées
7,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,202,Afrique du Sud,5511,Production,2513,Orge,2013,2013,Milliers de tonnes,268,S,Données standardisées
8,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,202,Afrique du Sud,5511,Production,2514,Maïs,2013,2013,Milliers de tonnes,11811,S,Données standardisées
9,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,202,Afrique du Sud,5511,Production,2515,Seigle,2013,2013,Milliers de tonnes,2,S,Données standardisées




📂 FICHIER : fr_population
📏 Dimensions : 175 lignes, 14 colonnes
------------------------------


,Code Domaine,Domaine,Code zone,Zone,Code Élément,Élément,Code Produit,Produit,Code année,Année,Unité,Valeur,Symbole,Description du Symbole
0,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,511,Population totale,2501,Population,2013,2013,1000 personnes,30552,NaN,Donnée officielle
1,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,202,Afrique du Sud,511,Population totale,2501,Population,2013,2013,1000 personnes,52776,NaN,Donnée officielle
2,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,3,Albanie,511,Population totale,2501,Population,2013,2013,1000 personnes,3173,NaN,Donnée officielle
3,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,4,Algérie,511,Population totale,2501,Population,2013,2013,1000 personnes,39208,NaN,Donnée officielle
4,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,79,Allemagne,511,Population totale,2501,Population,2013,2013,1000 personnes,82727,NaN,Donnée officielle
5,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,7,Angola,511,Population totale,2501,Population,2013,2013,1000 personnes,21472,NaN,Donnée officielle
6,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,8,Antigua-et-Barbuda,511,Population totale,2501,Population,2013,2013,1000 personnes,90,NaN,Donnée officielle
7,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,194,Arabie saoudite,511,Population totale,2501,Population,2013,2013,1000 personnes,28829,NaN,Donnée officielle
8,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,9,Argentine,511,Population totale,2501,Population,2013,2013,1000 personnes,41446,NaN,Donnée officielle
9,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,1,Arménie,511,Population totale,2501,Population,2013,2013,1000 personnes,2977,NaN,Donnée officielle




📂 FICHIER : fr_sousalimentation
📏 Dimensions : 1020 lignes, 15 colonnes
------------------------------


,Code Domaine,Domaine,Code zone,Zone,Code Élément,Élément,Code Produit,Produit,Code année,Année,Unité,Valeur,Symbole,Description du Symbole,Note
0,FS,Données de la sécurité alimentaire,2,Afghanistan,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20122014,2012-2014,millions,7.9,F,Estimation FAO,NaN
1,FS,Données de la sécurité alimentaire,2,Afghanistan,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20132015,2013-2015,millions,8.8,F,Estimation FAO,NaN
2,FS,Données de la sécurité alimentaire,2,Afghanistan,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20142016,2014-2016,millions,9.6,F,Estimation FAO,NaN
3,FS,Données de la sécurité alimentaire,2,Afghanistan,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20152017,2015-2017,millions,10.2,F,Estimation FAO,NaN
4,FS,Données de la sécurité alimentaire,2,Afghanistan,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20162018,2016-2018,millions,10.6,F,Estimation FAO,NaN
5,FS,Données de la sécurité alimentaire,202,Afrique du Sud,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20122014,2012-2014,millions,2.6,F,Estimation FAO,NaN
6,FS,Données de la sécurité alimentaire,202,Afrique du Sud,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20132015,2013-2015,millions,2.8,F,Estimation FAO,NaN
7,FS,Données de la sécurité alimentaire,202,Afrique du Sud,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20142016,2014-2016,millions,3.2,F,Estimation FAO,NaN
8,FS,Données de la sécurité alimentaire,202,Afrique du Sud,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20152017,2015-2017,millions,3.4,F,Estimation FAO,NaN
9,FS,Données de la sécurité alimentaire,202,Afrique du Sud,6132,Valeur,210011,Nombre de personnes sous-alimentées (millions)...,20162018,2016-2018,millions,3.5,F,Estimation FAO,NaN




📂 FICHIER : fr_vegetaux
📏 Dimensions : 104871 lignes, 14 colonnes
------------------------------


,Code Domaine,Domaine,Code zone,Zone,Code Élément,Élément,Code Produit,Produit,Code année,Année,Unité,Valeur,Symbole,Description du Symbole
0,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2511,Blé,2013,2013,Milliers de tonnes,5169.00,S,Données standardisées
1,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5611,Importations - Quantité,2511,Blé,2013,2013,Milliers de tonnes,1173.00,S,Données standardisées
2,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5072,Variation de stock,2511,Blé,2013,2013,Milliers de tonnes,-350.00,S,Données standardisées
3,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5301,Disponibilité intérieure,2511,Blé,2013,2013,Milliers de tonnes,5992.00,S,Données standardisées
4,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5527,Semences,2511,Blé,2013,2013,Milliers de tonnes,322.00,S,Données standardisées
5,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5123,Pertes,2511,Blé,2013,2013,Milliers de tonnes,775.00,S,Données standardisées
6,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5142,Nourriture,2511,Blé,2013,2013,Milliers de tonnes,4895.00,S,Données standardisées
7,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,645,Disponibilité alimentaire en quantité (kg/pers...,2511,Blé,2013,2013,kg,160.23,Fc,Donnée calculée
8,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,664,Disponibilité alimentaire (Kcal/personne/jour),2511,Blé,2013,2013,Kcal/personne/jour,1369.00,Fc,Donnée calculée
9,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,674,Disponibilité de protéines en quantité (g/pers...,2511,Blé,2013,2013,g/personne/jour,36.91,Fc,Donnée calculée


In [193]:
print("🔍 Rapport de Qualité des Données (Taux de complétude)\n")
print("-" * 60)

# On parcourt notre dictionnaire de DataFrames
for name, df in all_dataframes.items():
    total_rows = len(df)
    
    print(f"📊 FICHIER : {name} ({total_rows} lignes)")
    print("-" * 30)
    
    # On calcule les valeurs manquantes pour chaque colonne
    null_counts = df.isnull().sum()
    
    for col in null_counts.index:
        null_count = null_counts[col]
        
        if total_rows > 0:
            # Calcul du pourcentage de données manquantes
            pct_missing = (null_count / total_rows) * 100
            
            # On prépare un affichage propre
            # Si le manque est faible, on affiche en vert (symboliquement)
            # Si le manque est élevé (> 5%), on l'alerte
            status = "✅ OK" if pct_missing < 5 else "⚠️ ALERTE"
            
            print(f"  [{status}] {col:20} : {null_count:>6} manquantes ({pct_missing:>5.2f}%)")
        else:
            print((f"  ❌ Fichier vide pour la colonne {col}"))
            
    print("\n" + "="*60 + "\n")

🔍 Rapport de Qualité des Données (Taux de complétude)

------------------------------------------------------------
📊 FICHIER : fr_animaux (37166 lignes)
------------------------------
  [✅ OK] Code Domaine         :      0 manquantes ( 0.00%)
  [✅ OK] Domaine              :      0 manquantes ( 0.00%)
  [✅ OK] Code zone            :      0 manquantes ( 0.00%)
  [✅ OK] Zone                 :      0 manquantes ( 0.00%)
  [✅ OK] Code Élément         :      0 manquantes ( 0.00%)
  [✅ OK] Élément              :      0 manquantes ( 0.00%)
  [✅ OK] Code Produit         :      0 manquantes ( 0.00%)
  [✅ OK] Produit              :      0 manquantes ( 0.00%)
  [✅ OK] Code année           :      0 manquantes ( 0.00%)
  [✅ OK] Année                :      0 manquantes ( 0.00%)
  [✅ OK] Unité                :      0 manquantes ( 0.00%)
  [✅ OK] Valeur               :      0 manquantes ( 0.00%)
  [✅ OK] Symbole              :      0 manquantes ( 0.00%)
  [✅ OK] Description du Symbole :      0 manquan

Certaines colonnes sont très peu remplies voire complètement vides.  
Je supprimerai les colonnes 'symbole' de fr_population et 'note' de fr_sousalimentation.  
j'investiguerai sur la colonne 'Valeur' de fr_sousalimentation pour juger de l'intéret de cette colonne qui présente 40% de valeurs manquantes.

In [194]:
print("🔍 Rapport de Qualité des Données (Doublons)\n")
print("-" * 60)

for name, df in all_dataframes.items():
    total_rows = len(df)
    
    # --- PARTIE 1 : Vérification des doublons parfaits ---
    duplicate_count = df.duplicated().sum()
    
    print(f"📊 FICHIER : {name} ({total_rows} lignes)")
    print(f"  [!] Doublons parfaits trouvés : {duplicate_count}")

🔍 Rapport de Qualité des Données (Doublons)

------------------------------------------------------------
📊 FICHIER : fr_animaux (37166 lignes)
  [!] Doublons parfaits trouvés : 0
📊 FICHIER : fr_céréales (891 lignes)
  [!] Doublons parfaits trouvés : 0
📊 FICHIER : fr_population (175 lignes)
  [!] Doublons parfaits trouvés : 0
📊 FICHIER : fr_sousalimentation (1020 lignes)
  [!] Doublons parfaits trouvés : 0
📊 FICHIER : fr_vegetaux (104871 lignes)
  [!] Doublons parfaits trouvés : 0


Aucun doublon parfait  
En fonction des données regardées, nous analyserons plus en détail les éventuelles lignes doublées partiellement

In [195]:
print("🔍 Rapport de Qualité des Données (Dtypes)\n")
print("-" * 60)

for name, df in all_dataframes.items():
    print(f"📊 FICHIER : {name}")
    print(f"  Total colonnes : {len(df.columns)}")
    print("-" * 30)
    
    # On récupère les types et on crée un petit tableau récapitulatif
    dtype_info = []
    for col, dtype in df.dtypes.items():
        # On compte combien de valeurs ne sont PAS nulles pour voir la réalité du type
        non_null_count = df[col].count()
        dtype_info.append({
            "Colonne": col,
            "Type actuel": dtype,
  			"Valeurs non-nulles": non_null_count
        })
    
    # Création d'un DataFrame de diagnostic pour un affichage propre
    df_diag = pd.DataFrame(dtype_info)
    
    # On affiche le tableau de diagnostic
    display(df_diag)
    
    # Alerte spécifique : On cherche les colonnes numériques qui sont en type 'object'
    # (C'est souvent là que se cachent les erreurs de lecture)
    numeric_as_object = df_diag[(df_diag['Type actuel'] == 'object') & 
                                (df_diag['Colonne'].str.contains('Valeur|Année|Code', case=False, na=False))]
    
    if not numeric_as_object.empty:
        print(f"  ⚠️ ALERTE : Colonnes numériques potentiellement mal typées :")
        for _, row in numeric_as_object.iterrows():
            print(f"     👉 {row['Colonne']} est en type '{row['Type actuel']}'")
    else:
        print("  ✅ Aucun problème de type numérique détecté.")
        
    print("\n" + "="*60 + "\n")

🔍 Rapport de Qualité des Données (Dtypes)

------------------------------------------------------------
📊 FICHIER : fr_animaux
  Total colonnes : 14
------------------------------


,Colonne,Type actuel,Valeurs non-nulles
0,Code Domaine,str,37166
1,Domaine,str,37166
2,Code zone,int64,37166
3,Zone,str,37166
4,Code Élément,int64,37166
5,Élément,str,37166
6,Code Produit,int64,37166
7,Produit,str,37166
8,Code année,int64,37166
9,Année,int64,37166


  ✅ Aucun problème de type numérique détecté.


📊 FICHIER : fr_céréales
  Total colonnes : 14
------------------------------


,Colonne,Type actuel,Valeurs non-nulles
0,Code Domaine,str,891
1,Domaine,str,891
2,Code zone,int64,891
3,Zone,str,891
4,Code Élément,int64,891
5,Élément,str,891
6,Code Produit,int64,891
7,Produit,str,891
8,Code année,int64,891
9,Année,int64,891


  ✅ Aucun problème de type numérique détecté.


📊 FICHIER : fr_population
  Total colonnes : 14
------------------------------


,Colonne,Type actuel,Valeurs non-nulles
0,Code Domaine,str,175
1,Domaine,str,175
2,Code zone,int64,175
3,Zone,str,175
4,Code Élément,int64,175
5,Élément,str,175
6,Code Produit,int64,175
7,Produit,str,175
8,Code année,int64,175
9,Année,int64,175


  ✅ Aucun problème de type numérique détecté.


📊 FICHIER : fr_sousalimentation
  Total colonnes : 15
------------------------------


,Colonne,Type actuel,Valeurs non-nulles
0,Code Domaine,str,1020
1,Domaine,str,1020
2,Code zone,int64,1020
3,Zone,str,1020
4,Code Élément,int64,1020
5,Élément,str,1020
6,Code Produit,int64,1020
7,Produit,str,1020
8,Code année,int64,1020
9,Année,str,1020


  ✅ Aucun problème de type numérique détecté.


📊 FICHIER : fr_vegetaux
  Total colonnes : 14
------------------------------


,Colonne,Type actuel,Valeurs non-nulles
0,Code Domaine,str,104871
1,Domaine,str,104871
2,Code zone,int64,104871
3,Zone,str,104871
4,Code Élément,int64,104871
5,Élément,str,104871
6,Code Produit,int64,104871
7,Produit,str,104871
8,Code année,int64,104871
9,Année,int64,104871


  ✅ Aucun problème de type numérique détecté.




Les types des colonnes sont tous corrects

In [196]:
print("📊 Analyse descriptive des distributions\n")
print("=" * 60)

for name, df in all_dataframes.items():
    print(f"📂 FICHIER : {name}")
    print("-" * 30)
    
    # On sélectionne uniquement les colonnes numériques pour le describe
    df_numeric = df.select_dtypes(include=['number'])
    
    if df_numeric.empty:
        print("  ℹ️ Pas de colonnes numériques dans ce fichier.")
    else:
        # Le describe() avec les statistiques clés
        stats = df_numeric.describe().T  # .T pour transposer et avoir un tableau lisible
        
        # On ajoute la variance pour voir l'étalement des données
        stats['variance'] = df_numeric.var()
        
        # On affiche le résultat
        display(stats[['count', 'mean', 'std', 'variance', 'min', '50%', 'max']])
        
        print("\n  💡 Note d'analyse :")
        for col in df_numeric.columns:
            col_min = df_arg_min = df_numeric[col].min()
            col_max = df_numeric[col].max()
            col_mean = df_numeric[col].mean()
            
            # Si la moyenne est très loin de la médiane (50%), il y a un déséquilibre
            if abs(col_mean - df_numeric[col].median()) > (df_numeric[col].std() * 1):
                print(f"     ⚠️ Attention sur '{col}' : La moyenne et la médiane sont très décalées. Présence probable d'asymétrie.")

    print("\n" + "=" * 60 + "\n")

📊 Analyse descriptive des distributions

📂 FICHIER : fr_animaux
------------------------------


,count,mean,std,variance,min,50%,max
Code zone,37166.0,126.800839,74.767139,5.590125e+03,1.0,121.0,351.0
Code Élément,37166.0,3511.032718,2332.283315,5.439545e+06,645.0,5142.0,5911.0
Code Produit,37166.0,2758.421191,27.454449,7.537468e+02,2731.0,2762.0,2848.0
Code année,37166.0,2013.000000,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Année,37166.0,2013.000000,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Valeur,37166.0,148.291647,1933.728473,3.739306e+06,-1368.0,1.0,135600.0



  💡 Note d'analyse :


📂 FICHIER : fr_céréales
------------------------------


,count,mean,std,variance,min,50%,max
Code zone,891.0,130.005612,74.807040,5.596093e+03,1.0,129.0,276.0
Code Élément,891.0,5511.000000,0.000000,0.000000e+00,5511.0,5511.0,5511.0
Code Produit,891.0,2551.648709,96.152834,9.245367e+03,2511.0,2516.0,2805.0
Code année,891.0,2013.000000,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Année,891.0,2013.000000,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Valeur,891.0,2832.037037,16786.590126,2.817896e+08,0.0,95.0,353699.0



  💡 Note d'analyse :


📂 FICHIER : fr_population
------------------------------


,count,mean,std,variance,min,50%,max
Code zone,175.0,126.72,75.168519,5.650306e+03,1.0,121.0,351.0
Code Élément,175.0,511.00,0.000000,0.000000e+00,511.0,511.0,511.0
Code Produit,175.0,2501.00,0.000000,0.000000e+00,2501.0,2501.0,2501.0
Code année,175.0,2013.00,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Année,175.0,2013.00,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Valeur,175.0,48079.96,178632.742203,3.190966e+10,54.0,9413.0,1416667.0



  💡 Note d'analyse :


📂 FICHIER : fr_sousalimentation
------------------------------


,count,mean,std,variance,min,50%,max
Code zone,1020.0,1.292647e+02,76.615019,5.869861e+03,1.0,127.5,351.0
Code Élément,1020.0,6.132000e+03,0.000000,0.000000e+00,6132.0,6132.0,6132.0
Code Produit,1020.0,2.100110e+05,0.000000,0.000000e+00,210011.0,210011.0,210011.0
Code année,1020.0,2.014202e+07,14150.488052,2.002363e+08,20122014.0,20142016.0,20162018.0
Note,0.0,NaN,NaN,NaN,NaN,NaN,NaN



  💡 Note d'analyse :


📂 FICHIER : fr_vegetaux
------------------------------


,count,mean,std,variance,min,50%,max
Code zone,104871.0,127.694682,75.423255,5.688667e+03,1.0,121.0,351.0
Code Élément,104871.0,3728.248429,2276.474811,5.182338e+06,645.0,5142.0,5911.0
Code Produit,104871.0,2587.470998,61.107727,3.734154e+03,2511.0,2575.0,2899.0
Code année,104871.0,2013.000000,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Année,104871.0,2013.000000,0.000000,0.000000e+00,2013.0,2013.0,2013.0
Valeur,104871.0,321.271016,6395.039684,4.089653e+07,-39863.0,1.0,739267.0



  💡 Note d'analyse :




fr_vegetaux contient des nombres négatifs dans la colonne 'valeur', il faut s'assurer de la vraisemblance de cette donnée.  
L'année de référence des données est 2013, les données de la sous-nutrition comprennent plusieurs périodes, il faudra être vigilent à ne récuperer que l'année 2013 pour faire des corrélations.

In [197]:
all_dataframes['fr_vegetaux'][(all_dataframes['fr_vegetaux']['Code zone'] == 2)].head(20)

,Code Domaine,Domaine,Code zone,Zone,Code Élément,Élément,Code Produit,Produit,Code année,Année,Unité,Valeur,Symbole,Description du Symbole
0,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2511,Blé,2013,2013,Milliers de tonnes,5169.00,S,Données standardisées
1,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5611,Importations - Quantité,2511,Blé,2013,2013,Milliers de tonnes,1173.00,S,Données standardisées
2,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5072,Variation de stock,2511,Blé,2013,2013,Milliers de tonnes,-350.00,S,Données standardisées
3,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5301,Disponibilité intérieure,2511,Blé,2013,2013,Milliers de tonnes,5992.00,S,Données standardisées
4,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5527,Semences,2511,Blé,2013,2013,Milliers de tonnes,322.00,S,Données standardisées
5,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5123,Pertes,2511,Blé,2013,2013,Milliers de tonnes,775.00,S,Données standardisées
6,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5142,Nourriture,2511,Blé,2013,2013,Milliers de tonnes,4895.00,S,Données standardisées
7,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,645,Disponibilité alimentaire en quantité (kg/pers...,2511,Blé,2013,2013,kg,160.23,Fc,Donnée calculée
8,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,664,Disponibilité alimentaire (Kcal/personne/jour),2511,Blé,2013,2013,Kcal/personne/jour,1369.00,Fc,Donnée calculée
9,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,674,Disponibilité de protéines en quantité (g/pers...,2511,Blé,2013,2013,g/personne/jour,36.91,Fc,Donnée calculée


In [198]:
# j'essaie d'avoir l'intitulé complet d'une colonne
all_dataframes['fr_vegetaux'][(all_dataframes['fr_vegetaux']['Code zone'] == 2)]['Élément'].values

<StringArray>
[                                                   'Production',
                                       'Importations - Quantité',
                                            'Variation de stock',
                                      'Disponibilité intérieure',
                                                      'Semences',
                                                        'Pertes',
                                                    'Nourriture',
        'Disponibilité alimentaire en quantité (kg/personne/an)',
                'Disponibilité alimentaire (Kcal/personne/jour)',
      'Disponibilité de protéines en quantité (g/personne/jour)',
 ...
                                       'Importations - Quantité',
                                      'Disponibilité intérieure',
                                                    'Nourriture',
        'Disponibilité alimentaire en quantité (kg/personne/an)',
                'Disponibilité alimentaire (Kcal/personne

376 lignes de données par pays concernant l'alimentation végétale, ce n'est pas exploitables en l'état, nous ne pouvons pas faire de statistiques sur autant de données à la fois et certaines données se recoupent probablement.  
ici on cherche plutot un global d'aliments, ressources végétale pour la population d'un pays, pas le détail de calories contenues dans un type d'aliment.  
les données qui semblent intéressantes sont :
['Production' , 'Importations - Quantité', 'Disponibilité intérieure', 'Disponibilité alimentaire (Kcal/personne/jour)', 'Disponibilité alimentaire en quantité (kg/personne/an)'] sous forme de total par pays

In [199]:
# je regarde les valeurs uniques de la colonne 'Élément' pour vérifier qu'il n'y a pas de fautes
elements_uniques = all_dataframes['fr_vegetaux']['Élément'].unique()

print(f"Nombre d'éléments uniques : {len(elements_uniques)}")
print("Liste des éléments :\n", elements_uniques)

Nombre d'éléments uniques : 15
Liste des éléments :
 <StringArray>
[                                                   'Production',
                                       'Importations - Quantité',
                                            'Variation de stock',
                                      'Disponibilité intérieure',
                                                      'Semences',
                                                        'Pertes',
                                                    'Nourriture',
        'Disponibilité alimentaire en quantité (kg/personne/an)',
                'Disponibilité alimentaire (Kcal/personne/jour)',
      'Disponibilité de protéines en quantité (g/personne/jour)',
 'Disponibilité de matière grasse en quantité (g/personne/jour)',
                                         'Aliments pour animaux',
                                       'Exportations - Quantité',
                                                    'Traitement',
         

pas d'élément unique avec faute : OK  
  
Faisons la même chose avec le dataset fr_animaux

In [200]:
all_dataframes['fr_animaux'][(all_dataframes['fr_vegetaux']['Code zone'] == 2)].head(20)

C:\Users\johan\AppData\Local\Temp\ipykernel_19416\137660240.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  all_dataframes['fr_animaux'][(all_dataframes['fr_vegetaux']['Code zone'] == 2)].head(20)


,Code Domaine,Domaine,Code zone,Zone,Code Élément,Élément,Code Produit,Produit,Code année,Année,Unité,Valeur,Symbole,Description du Symbole
0,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2731,Viande de Bovins,2013,2013,Milliers de tonnes,134.00,S,Données standardisées
1,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5611,Importations - Quantité,2731,Viande de Bovins,2013,2013,Milliers de tonnes,6.00,S,Données standardisées
2,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5301,Disponibilité intérieure,2731,Viande de Bovins,2013,2013,Milliers de tonnes,140.00,S,Données standardisées
3,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5142,Nourriture,2731,Viande de Bovins,2013,2013,Milliers de tonnes,140.00,S,Données standardisées
4,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,645,Disponibilité alimentaire en quantité (kg/pers...,2731,Viande de Bovins,2013,2013,kg,4.59,Fc,Donnée calculée
5,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,664,Disponibilité alimentaire (Kcal/personne/jour),2731,Viande de Bovins,2013,2013,Kcal/personne/jour,27.00,Fc,Donnée calculée
6,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,674,Disponibilité de protéines en quantité (g/pers...,2731,Viande de Bovins,2013,2013,g/personne/jour,1.89,Fc,Donnée calculée
7,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,684,Disponibilité de matière grasse en quantité (g...,2731,Viande de Bovins,2013,2013,g/personne/jour,2.10,Fc,Donnée calculée
8,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5511,Production,2732,Viande d'Ovins/Caprins,2013,2013,Milliers de tonnes,150.00,S,Données standardisées
9,FBSH,Bilans Alimentaire (Ancienne méthodologie et p...,2,Afghanistan,5301,Disponibilité intérieure,2732,Viande d'Ovins/Caprins,2013,2013,Milliers de tonnes,150.00,S,Données standardisées


In [201]:
# je regarde les valeurs uniques de la colonne 'Élément' pour vérifier qu'il n'y a pas de fautes
elements_uniques = all_dataframes['fr_animaux']['Élément'].unique()

print(f"Nombre d'éléments uniques : {len(elements_uniques)}")
print("Liste des éléments :\n", elements_uniques)

Nombre d'éléments uniques : 15
Liste des éléments :
 <StringArray>
[                                                   'Production',
                                       'Importations - Quantité',
                                      'Disponibilité intérieure',
                                                    'Nourriture',
        'Disponibilité alimentaire en quantité (kg/personne/an)',
                'Disponibilité alimentaire (Kcal/personne/jour)',
      'Disponibilité de protéines en quantité (g/personne/jour)',
 'Disponibilité de matière grasse en quantité (g/personne/jour)',
                                                      'Semences',
                                                        'Pertes',
                                       'Exportations - Quantité',
                                         'Aliments pour animaux',
                         'Autres utilisations (non alimentaire)',
                                            'Variation de stock',
         

idem que fr_vegetaux : OK, pas d'anomalie  
fr_cereales est une sous_partie de fr_vegetaux : je ne le traite pas

In [202]:
# je vérifie l'existence des 15 lignes distinctes d'Element pour chaque 'Code Produit' dans fr_animaux et fr_vegetaux

print("🔍 Diagnostic détaillé des éléments manquants\n")
print("=" * 60)

def diagnose_missing_elements(df, df_name):
    if 'Élément' not in df.columns or 'Produit' not in df.columns or 'Code Produit' not in df.columns:
        return

    # On récupère la liste complète théorique des 15 éléments attendus
    total_expected = df['Élément'].nunique() 
    print(f"📊 Analyse sur {df_name} ({total_expected} éléments théoriques)")
    
    # On groupe par 'Code Produit' et on regarde quels éléments sont présents
    for prod_code, group_data in df.groupby('Code Produit'):
        
        # On récupère le nom du produit (on prend juste le premier trouvé)
        prod_name = group_data['Produit'].iloc[0]
        
        # On compare les éléments présents avec tous les éléments possibles du dataset
        present_elements = set(group_data['Élément'])
        all_elements = set(df['Élément'])
        
        missing_elements = list(all_elements - present_elements)
        
        if missing_elements:
            print(f"⚠️ Produit : {prod_name} (Code: {prod_code})")
            for el in missing_elements:
                print(f"   -> Élément manquant : {el}")
            print("-" * 40)

# Lancement du diagnostic sur tes fichiers principaux
diagnose_missing_elements(all_dataframes['fr_vegetaux'], 'fr_vegetaux.csv')
diagnose_missing_elements(all_dataframes['fr_animaux'], 'fr_animaux.csv')

🔍 Diagnostic détaillé des éléments manquants

📊 Analyse sur fr_vegetaux.csv (15 éléments théoriques)
⚠️ Produit : Ignames (Code: 2535)
   -> Élément manquant : Traitement
----------------------------------------
⚠️ Produit : Sucre, betterave (Code: 2537)
   -> Élément manquant : Semences
----------------------------------------
⚠️ Produit : Sucre non centrifugé (Code: 2541)
   -> Élément manquant : Semences
----------------------------------------
⚠️ Produit : Sucre Eq Brut (Code: 2542)
   -> Élément manquant : Disponibilité de matière grasse en quantité (g/personne/jour)
   -> Élément manquant : Disponibilité de protéines en quantité (g/personne/jour)
   -> Élément manquant : Semences
----------------------------------------
⚠️ Produit : Haricots (Code: 2546)
   -> Élément manquant : Traitement
----------------------------------------
⚠️ Produit : Noix (Code: 2551)
   -> Élément manquant : Aliments pour animaux
----------------------------------------
⚠️ Produit : Palmistes (Code: 256

In [203]:
print("📊 AUDIT RÉALISTE : Pourcentage de présence par Élément\n")
print("=" * 70)

def audit_coverage_pct(df, df_name):
    # 1. Le nombre total de pays uniques présents dans le fichier (ta base de comparaison)
    total_countries = df['Zone'].nunique()
    
    print(f"📂 Fichier : {df_name} | Total de pays dans le dataset : {total_countries}")
    print("-" * 70)
    
    stats = []
    
    for element in df['Élément'].unique():
        # On filtre l'élément et on enlève les valeurs nulles sur la 'Valeur'
        subset = df[df['Élément'] == element].dropna(subset=['Valeur'])
        
        # Combien de pays ont réellement fourni une valeur pour cet élément ?
        n_countries_present = subset['Zone'].nunique()
        
        # Calcul du pourcentage
        pct_presence = (n_countries_present / total_countries) * 100
        
        stats.append({
            'Élément': element,
            'Pays présents': n_countries_present,
            'Total Pays théoriques': total_countries,
            '% de Présence (%)': round(pct_presence, 2)
        })

    # Affichage trié du plus complet au moins complet
    df_res = pd.DataFrame(stats).sort_values(by='% de Présence (%)', ascending=False)
    
    display(df_res)
    print("\n" + "="*70 + "\n")

# Lancement sur tes fichiers
audit_coverage_pct(all_dataframes['fr_vegetaux'], 'fr_vegetaux.csv')
audit_coverage_pct(all_dataframes['fr_animaux'], 'fr_animaux.csv')


print("🌍 VUE MACROSCOPIQUE : Pourcentage de présence Global\n")
print("=" * 70)

def audit_global_presence_pct(df, df_name):
    # 1. Le total théorique possible (l'espace vide si tout était rempli)
    n_produits = df['Code Produit'].nunique()
    n_pays = df['Zone'].nunique()
    
    # On suppose qu'il y a une année par ligne pour l'estimation du max
    # (Si tu as plusieurs années, le dénominateur sera plus grand)
    total_possible_lines = n_produits * n_pays 
    
    print(f"📂 Fichier : {df_name}")
    print(f"   -> Dimensions théoriques maximales (Produits x Pays x Années) : {total_possible_lines}")
    print("-" * 70)
    
    stats = []
    
    for element in df['Élément'].unique():
        # On compte les lignes réelles pour cet élément
        real_lines = len(df[df['Élément'] == element])
        
        # Calcul du pourcentage de remplissage du "potentiel" total
        global_pct = (real_lines / total_possible_lines) * 100
        
        stats.append({
            'Élément': element,
            'Lignes réelles (Total Mondial)': real_lines,
            '% de Présence Global (%)': round(global_pct, 2)
        })

    # Affichage trié
    df_res = pd.DataFrame(stats).sort_values(by='% de Présence Global (%)', ascending=False)
    
    display(df_res)
    print("\n" + "="*70 + "\n")

# Lancement sur tes fichiers
audit_global_presence_pct(all_dataframes['fr_vegetaux'], 'fr_vegetaux.csv')
audit_global_presence_pct(all_dataframes['fr_animaux'], 'fr_animaux.csv')

📊 AUDIT RÉALISTE : Pourcentage de présence par Élément

📂 Fichier : fr_vegetaux.csv | Total de pays dans le dataset : 175
----------------------------------------------------------------------


,Élément,Pays présents,Total Pays théoriques,% de Présence (%)
0,Production,175,175,100.00
1,Importations - Quantité,175,175,100.00
2,Variation de stock,175,175,100.00
3,Disponibilité intérieure,175,175,100.00
6,Nourriture,175,175,100.00
8,Disponibilité alimentaire (Kcal/personne/jour),175,175,100.00
7,Disponibilité alimentaire en quantité (kg/pers...,175,175,100.00
11,Aliments pour animaux,175,175,100.00
12,Exportations - Quantité,175,175,100.00
9,Disponibilité de protéines en quantité (g/pers...,175,175,100.00




📂 Fichier : fr_animaux.csv | Total de pays dans le dataset : 175
----------------------------------------------------------------------


,Élément,Pays présents,Total Pays théoriques,% de Présence (%)
0,Production,175,175,100.00
1,Importations - Quantité,175,175,100.00
2,Disponibilité intérieure,175,175,100.00
3,Nourriture,175,175,100.00
4,Disponibilité alimentaire en quantité (kg/pers...,175,175,100.00
5,Disponibilité alimentaire (Kcal/personne/jour),175,175,100.00
6,Disponibilité de protéines en quantité (g/pers...,175,175,100.00
7,Disponibilité de matière grasse en quantité (g...,175,175,100.00
10,Exportations - Quantité,175,175,100.00
12,Autres utilisations (non alimentaire),175,175,100.00




🌍 VUE MACROSCOPIQUE : Pourcentage de présence Global

📂 Fichier : fr_vegetaux.csv
   -> Dimensions théoriques maximales (Produits x Pays x Années) : 13125
----------------------------------------------------------------------


,Élément,Lignes réelles (Total Mondial),% de Présence Global (%)
3,Disponibilité intérieure,11747,89.50
1,Importations - Quantité,11308,86.16
8,Disponibilité alimentaire (Kcal/personne/jour),10607,80.82
7,Disponibilité alimentaire en quantité (kg/pers...,10381,79.09
6,Nourriture,10381,79.09
12,Exportations - Quantité,9299,70.85
10,Disponibilité de matière grasse en quantité (g...,8156,62.14
9,Disponibilité de protéines en quantité (g/pers...,7944,60.53
0,Production,6312,48.09
2,Variation de stock,5446,41.49




📂 Fichier : fr_animaux.csv
   -> Dimensions théoriques maximales (Produits x Pays x Années) : 4025
----------------------------------------------------------------------


,Élément,Lignes réelles (Total Mondial),% de Présence Global (%)
2,Disponibilité intérieure,3731,92.70
4,Disponibilité alimentaire en quantité (kg/pers...,3726,92.57
3,Nourriture,3726,92.57
5,Disponibilité alimentaire (Kcal/personne/jour),3726,92.57
7,Disponibilité de matière grasse en quantité (g...,3721,92.45
6,Disponibilité de protéines en quantité (g/pers...,3697,91.85
1,Importations - Quantité,3638,90.39
10,Exportations - Quantité,3021,75.06
0,Production,2959,73.52
12,Autres utilisations (non alimentaire),2397,59.55


Les 175 pays sont bien couverts par les données essentielles qui nous intéressent.  
Tous les produits ne sont pas complètement renseignés mais des lignes manquantes signifient dans certains cas 0 (ex du thé avec les matières grasse, du poivre avec l'alimentation pour animaux) ou certains produits sont absents de l'alimentation de certains pays.  
Ici une correction serait impossible, fausserait les données.  


In [204]:
# Je vérifie l'homogénéité des unités

print("🔍 AUDIT RAPIDE : Unités associées aux Éléments\n")
print("=" * 70)

def audit_unit_by_element(df, df_name):
    if 'Unité' not in df.columns or 'Élément' not in df.columns:
        print(f"❌ Colonnes manquantes dans {df_name}")
        return

    # Dictionnaire pour grouper : Clé = Élément, Valeur = Ensemble d'unités
    elements_to_units = {}

    for index, row in df.iterrows():
        el = row['Élément']
        unit = row['Unité']
        
        if el not in elements_to_units:
            elements_to_units[el] = set()
        
        elements_to_units[el].add(unit)

    # Affichage structuré
    print(f"📂 Fichier : {df_name}")
    print("-" * 70)
    
    for el, units in sorted(elements_to_units.items()):
        units_list = ", ".join(sorted(units))
        print(f"   📌 {el:25} -> Unités : [{units_list}]")

    print("\n" + "="*70 + "\n")

# Lancement sur tes fichiers
audit_unit_by_element(all_dataframes['fr_vegetaux'], 'fr_vegetaux.csv')
audit_unit_by_element(all_dataframes['fr_animaux'], 'fr_animaux.csv')

🔍 AUDIT RAPIDE : Unités associées aux Éléments

📂 Fichier : fr_vegetaux.csv
----------------------------------------------------------------------
   📌 Aliments pour animaux     -> Unités : [Milliers de tonnes]
   📌 Autres utilisations (non alimentaire) -> Unités : [Milliers de tonnes]
   📌 Disponibilité alimentaire (Kcal/personne/jour) -> Unités : [Kcal/personne/jour]
   📌 Disponibilité alimentaire en quantité (kg/personne/an) -> Unités : [kg]
   📌 Disponibilité de matière grasse en quantité (g/personne/jour) -> Unités : [g/personne/jour]
   📌 Disponibilité de protéines en quantité (g/personne/jour) -> Unités : [g/personne/jour]
   📌 Disponibilité intérieure  -> Unités : [Milliers de tonnes]
   📌 Exportations - Quantité   -> Unités : [Milliers de tonnes]
   📌 Importations - Quantité   -> Unités : [Milliers de tonnes]
   📌 Nourriture                -> Unités : [Milliers de tonnes]
   📌 Pertes                    -> Unités : [Milliers de tonnes]
   📌 Production                -> Unités :

J'ai bien une unité par type d'élément mesuré et les unités sont compatibles entre les 2 dataframes

In [205]:
"""
Script de contrôle de cohérence des données avant fusion.
Vérifie que chaque code correspond toujours au même libellé dans tous les fichiers.
"""

def check_consistency(all_dataframes, key_pairs, file_list):
    """
    Vérifie la cohérence entre les paires de colonnes spécifiées.
    
    Args:
        all_dataframes: Dictionnaire des DataFrames
        key_pairs: Liste de tuples ('colonne_code', 'colonne_libelle')
        file_list: Liste des fichiers à vérifier
    
    Returns:
        Résultats avec erreurs par fichier
    """
    # Structure pour stocker les liens observés : {code: set(libellés)}
    global_links = {}
    errors_per_file = {}

    for file_name in file_list:
        if file_name not in all_dataframes:
            if file_name not in errors_per_file:
                errors_per_file[file_name] = []
            errors_per_file[file_name].append("Fichier introuvable dans 'all_dataframes'.")
            continue

        df = all_dataframes[file_name]
        code_col, label_col = key_pairs[0]  # On suppose une seule paire pour simplifier

        if code_col not in df.columns or label_col not in df.columns:
            if file_name not in errors_per_file:
                errors_per_file[file_name] = []
            errors_per_file[file_name].append(f"Colonnes manquantes : {code_col}, {label_col}")
            continue

        # Extraction des paires uniques
        unique_pairs = df[[code_col, label_col]].drop_duplicates()

        for index, row in unique_pairs.iterrows():
            code = row[code_col]
            label = row[label_col]

            if code not in global_links:
                global_links[code] = set()
                global_links[code].add(label)
            else:
                existing_labels = global_links[code]
                if label not in existing_labels:
                    if file_name not in errors_per_file:
                        errors_per_file[file_name] = []
                    errors_per_file[file_name].append(
                        f"Conflit sur {code_col} '{code}' : '{list(existing_labels)[0]}' (déjà connu) vs '{label}' (dans {file_name})"
                    )
    return errors_per_file

print("🔍 AUDIT GLOBAL : Cohérence des clés de fusion")
print("=" * 80)

# Fichiers concernés pour chaque type de clé
geo_files = ['fr_vegetaux', 'fr_animaux', 'fr_population', 'fr_sousalimentation']
prod_files = ['fr_vegetaux', 'fr_animaux']
element_files = ['fr_vegetaux', 'fr_animaux']  # Supposons qu'ils contiennent Element

# Vérification des clés géographiques
print("\n📋 Cohérence Zone/Code zone")
zone_errors = check_consistency(
        all_dataframes={k: df for k, df in all_dataframes.items() if k in geo_files},
        key_pairs=[('Code zone', 'Zone')],
        file_list=geo_files
    )
    
# Vérification des clés de produits
print("\n📋 Cohérence Code Produit/Produit")
prod_errors = check_consistency(
        all_dataframes={k: df for k, df in all_dataframes.items() if k in prod_files},
        key_pairs=[('Code Produit', 'Produit')],
        file_list=prod_files
    )

# Vérification des clés d'éléments (orthographe exact : Code Élément / Élément)
print("\n📋 Cohérence Code Élément/Élément")
element_errors = check_consistency(
        all_dataframes={k: df for k, df in all_dataframes.items() if k in element_files},
        key_pairs=[('Code Élément', 'Élément')],
        file_list=element_files
    )

# Affichage des résultats
print("\n" + "=" * 80)
print("📊 RÉSUMÉ DES ERREURS")
print("=" * 80)

for name, errors in [("Zone", zone_errors), ("Produit", prod_errors), ("Élément", element_errors)]:
    if any(errors.values()):
        print(f"\n❌ Erreurs dans la cohérence {name}:")
        for file_name, file_errors in errors.items():
            if file_errors:
                print(f"   {file_name}:")
                for err in file_errors:
                    print(f"     - {err}")
    else:
        print(f"\n✅ Aucune erreur détectée pour la cohérence {name}")

print("\n" + "=" * 80)
print("Fin de l'audit")


🔍 AUDIT GLOBAL : Cohérence des clés de fusion

📋 Cohérence Zone/Code zone

📋 Cohérence Code Produit/Produit

📋 Cohérence Code Élément/Élément

📊 RÉSUMÉ DES ERREURS

✅ Aucune erreur détectée pour la cohérence Zone

✅ Aucune erreur détectée pour la cohérence Produit

✅ Aucune erreur détectée pour la cohérence Élément

Fin de l'audit


les code 'item' et leur 'item' respectif sont tous correctement corrélés, aucun doublon, aucune erreur  
=> on peut faire des fusions avec ces clés sans crainte d'erreur

## Nettoyage

In [206]:
# je supprime fr_cereales qui est inclus dans fr_vegetaux
# Suppression du fichier fr_cereales du dictionnaire global
if 'fr_céréales' in all_dataframes:
    del all_dataframes['fr_céréales']
    print("✅ fr_céréales supprimé du dictionnaire.")
else:
    print("ℹ️ Le fichier fr_céréales n'était pas dans le dictionnaire.")

print(f"\n📂 Fichiers restants dans ton projet : {list(all_dataframes.keys())}")


# On nettoie les colonnes inutiles
cols_to_drop = [
    'Symbole', 
    'Description du Symbole', 
    'Domaine', 
    'Code Domaine', 
    'Code année'
]

# On parcourt chaque fichier du dictionnaire, peu importe son nom ou sa structure
for name, df in all_dataframes.items():
    
    # On identifie les colonnes à supprimer qui sont réellement présentes dans ce fichier
    cols_presentes = [col for col in cols_to_drop if col in df.columns]
    
    if cols_presentes:
        # Suppression et mise à jour du dictionnaire
        all_dataframes[name] = df.drop(columns=cols_presentes)
        print(f"✅ {name} : Nettoyé ({len(cols_presentes)} colonnes supprimées).")
    else:
        # Optionnel : on peut masquer ce message si tu préfères un silence total
        pass

print("\n✨ Tous les fichiers du dictionnaire ont été passés au crible.")
print(f"📂 État final des fichiers : {list(all_dataframes.keys())}")


✅ fr_céréales supprimé du dictionnaire.

📂 Fichiers restants dans ton projet : ['fr_animaux', 'fr_population', 'fr_sousalimentation', 'fr_vegetaux']
✅ fr_animaux : Nettoyé (5 colonnes supprimées).
✅ fr_population : Nettoyé (5 colonnes supprimées).
✅ fr_sousalimentation : Nettoyé (5 colonnes supprimées).
✅ fr_vegetaux : Nettoyé (5 colonnes supprimées).

✨ Tous les fichiers du dictionnaire ont été passés au crible.
📂 État final des fichiers : ['fr_animaux', 'fr_population', 'fr_sousalimentation', 'fr_vegetaux']


## Analyse exploratoire

Ce qui nous intéresse est la quantité de nourriture, nutriments, disponibilité globale par pays et non pas la répartition, les coutumes alimentaires.
Je vais donc additionner les données alimentaires pour chaque catégorie

In [207]:
# Création d'une fonction pour aggréger les données et retourner les nouveaux DataFrames
def create_aggregated_dataframe(df, df_name):
    """
    Crée un DataFrame agrégé à partir d'un DataFrame existant.
    
    Args:
        df: DataFrame à agréger
        df_name: Nom du DataFrame (pour affichage)
    
    Returns:
        DataFrame agrégé
    """
    import pandas as pd
    
    print(f"🔍 Création de DataFrame agrégé pour {df_name}...")
    
    # Vérification que les colonnes nécessaires existent
    required_cols = ['Code zone', 'Élément', 'Valeur', 'Produit', 'Unité']
    if not all(col in df.columns for col in required_cols):
        print(f"❌ Colonnes manquantes dans {df_name}. Colonnes requises : {required_cols}")
        return df
    
    # Affichage des statistiques avant aggrégation
    print(f"   Données originales : {len(df)} lignes")
    
    # Création d'un nouveau DataFrame avec les agrégations
    # Grouper par Code zone et Élément, sommer les Valeurs 
    # Pour Unité, on prend la première valeur (la plus représentative) pour chaque groupe
    aggregated_df = df.groupby(['Code zone', 'Élément'], as_index=False).agg({
        'Valeur': 'sum',
        'Produit': 'first',  # On garde le premier produit (sera 'Total' après)
        'Unité': 'first'     # On garde la première unité pour chaque groupe
    })
    
    # Remplacer tous les produits par 'Total' 
    aggregated_df['Produit'] = 'Total'
    
    # Conserver les colonnes originales nécessaires
    # Ajouter les colonnes qui n'existaient pas dans le groupby mais qui sont dans l'original
    original_columns = df.columns.tolist()
    
    # Pour chaque colonne de l'original, si elle n'existe pas dans le nouveau df, on la remplit
    for col in original_columns:
        if col not in aggregated_df.columns and col != 'Code Produit':
            if col == 'Zone':
                # Récupérer les zones correspondantes à partir des Code zone
                zone_mapping = df[['Code zone', 'Zone']].drop_duplicates()
                aggregated_df = aggregated_df.merge(zone_mapping, on='Code zone', how='left')
            elif col == 'Code Élément':
                # Pour Code Élément, on peut garder la première valeur ou générer une nouvelle valeur
                # Ici, on garde la première valeur pour chaque groupe
                code_element_mapping = df.groupby(['Code zone', 'Élément'])['Code Élément'].first().reset_index()
                aggregated_df = aggregated_df.merge(code_element_mapping, on=['Code zone', 'Élément'], how='left')
            elif col == 'Année':
                # Pour Année, garder la première valeur
                année_mapping = df.groupby(['Code zone', 'Élément'])['Année'].first().reset_index()
                aggregated_df = aggregated_df.merge(année_mapping, on=['Code zone', 'Élément'], how='left')
            else:
                # Pour les autres colonnes, on garde la première valeur
                mapping = df.groupby(['Code zone', 'Élément'])[col].first().reset_index()
                aggregated_df = aggregated_df.merge(mapping, on=['Code zone', 'Élément'], how='left')
    
    # Supprimer la colonne 'Code Produit' qui n'est plus pertinente
    if 'Code Produit' in aggregated_df.columns:
        aggregated_df = aggregated_df.drop('Code Produit', axis=1)
    
    # Réorganiser les colonnes pour garder la même structure que l'original
    existing_cols = [col for col in original_columns if col != 'Code Produit' and col in aggregated_df.columns]
    aggregated_df = aggregated_df[existing_cols]
    
    print(f"   ✅ DataFrame agrégé créé : {aggregated_df.shape}")
    print(f"   ✅ Colonnes : {list(aggregated_df.columns)}")
    
    return aggregated_df

# Création des DataFrames agrégés
print("📊 Création des DataFrames agrégés")
print("=" * 50)

# Agrégation de fr_vegetaux
if 'fr_vegetaux' in all_dataframes:
    aggregated_vegetaux = create_aggregated_dataframe(all_dataframes['fr_vegetaux'], 'fr_vegetaux')
    print("\n📋 Aperçu du DataFrame agrégé fr_vegetaux :")
    display(aggregated_vegetaux.head())

# Agrégation de fr_animaux  
if 'fr_animaux' in all_dataframes:
    aggregated_animaux = create_aggregated_dataframe(all_dataframes['fr_animaux'], 'fr_animaux')
    print("\n📋 Aperçu du DataFrame agrégé fr_animaux :")
    display(aggregated_animaux.head())

📊 Création des DataFrames agrégés
🔍 Création de DataFrame agrégé pour fr_vegetaux...
   Données originales : 104871 lignes
   ✅ DataFrame agrégé créé : (2619, 8)
   ✅ Colonnes : ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'Valeur']

📋 Aperçu du DataFrame agrégé fr_vegetaux :


,Code zone,Zone,Code Élément,Élément,Produit,Année,Unité,Valeur
0,1,Arménie,5521,Aliments pour animaux,Total,2013,Milliers de tonnes,601.00
1,1,Arménie,5154,Autres utilisations (non alimentaire),Total,2013,Milliers de tonnes,235.00
2,1,Arménie,664,Disponibilité alimentaire (Kcal/personne/jour),Total,2013,Kcal/personne/jour,2140.00
3,1,Arménie,645,Disponibilité alimentaire en quantité (kg/pers...,Total,2013,kg,704.69
4,1,Arménie,684,Disponibilité de matière grasse en quantité (g...,Total,2013,g/personne/jour,39.00


🔍 Création de DataFrame agrégé pour fr_animaux...
   Données originales : 37166 lignes
   ✅ DataFrame agrégé créé : (2517, 8)
   ✅ Colonnes : ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'Valeur']

📋 Aperçu du DataFrame agrégé fr_animaux :


,Code zone,Zone,Code Élément,Élément,Produit,Année,Unité,Valeur
0,1,Arménie,5521,Aliments pour animaux,Total,2013,Milliers de tonnes,56.00
1,1,Arménie,5154,Autres utilisations (non alimentaire),Total,2013,Milliers de tonnes,7.00
2,1,Arménie,664,Disponibilité alimentaire (Kcal/personne/jour),Total,2013,Kcal/personne/jour,784.00
3,1,Arménie,645,Disponibilité alimentaire en quantité (kg/pers...,Total,2013,kg,280.99
4,1,Arménie,684,Disponibilité de matière grasse en quantité (g...,Total,2013,g/personne/jour,57.63


In [208]:
## Fusion des DataFrames agrégés

def merge_aggregated_dataframes(df1_name=aggregated_vegetaux, df2_name=aggregated_animaux):
    """
    Fusionne deux DataFrames agrégés avec une jointure outer sur Code zone et Code Élément.
    Additionne les valeurs lorsque les deux clés correspondent.
    
    Args:
        df1_name: Nom du premier DataFrame
        df2_name: Nom du second DataFrame
    
    Returns:
        DataFrame fusionné
    """
    df1 = df1_name
    df2 = df2_name    
    print(f"🔍 Fusion")
    print(f"   Taille df1 : {df1.shape}")
    print(f"   Taille df2 : {df2.shape}")
    
    # Vérification des colonnes
    print(f"   Colonnes df1 : {list(df1.columns)}")
    print(f"   Colonnes df2 : {list(df2.columns)}")
    
    # Fusion avec outer join sur Code zone et Code Élément
    merged_df = pd.merge(df1, df2, on=['Code zone', 'Code Élément'], how='outer', suffixes=('_1', '_2'))
    
    # Addition des valeurs lorsque les deux clés correspondent
    # Pour les colonnes communes (sauf les clés), on additionne les valeurs
    value_cols = ['Valeur']
    
    for col in value_cols:
        if f"{col}_1" in merged_df.columns and f"{col}_2" in merged_df.columns:
            # Addition des deux colonnes
            merged_df[col] = merged_df[f"{col}_1"] + merged_df[f"{col}_2"]
            # Suppression des colonnes temporaires
            merged_df = merged_df.drop(columns=[f"{col}_1", f"{col}_2"])
    
    # Pour les autres colonnes, on garde les valeurs du premier DataFrame si elles existent
    # ou celles du second si les premières sont vides
    common_cols = ['Zone', 'Élément', 'Produit', 'Année', 'Unité']
    
    for col in common_cols:
        if f"{col}_1" in merged_df.columns and f"{col}_2" in merged_df.columns:
            # On garde les valeurs du premier DataFrame, et on remplit avec le second si vide
            merged_df[col] = merged_df[f"{col}_1"].fillna(merged_df[f"{col}_2"])
            merged_df = merged_df.drop(columns=[f"{col}_1", f"{col}_2"])
    
    # Réorganiser les colonnes dans l'ordre original
    original_cols = ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'Valeur']
    existing_cols = [col for col in original_cols if col in merged_df.columns]
    merged_df = merged_df[existing_cols]
    
    print(f"   ✅ Fusion terminée : {merged_df.shape}")
    print(f"   ✅ Colonnes finales : {list(merged_df.columns)}")
    
    return merged_df

# Exécution de la fusion
print("📊 Fusion des DataFrames agrégés")
print("=" * 50)

# Création du DataFrame fusionné
merged_df = merge_aggregated_dataframes(aggregated_vegetaux, aggregated_animaux)

print("\n📋 Aperçu du DataFrame fusionné :")
display(merged_df.head(10))
    
print(f"\n📊 Statistiques du DataFrame fusionné :")
print(f"   Lignes : {len(merged_df)}")
print(f"   Colonnes : {len(merged_df.columns)}")
print(f"   Pays uniques : {merged_df['Code zone'].nunique()}")
print(f"   Éléments uniques : {merged_df['Code Élément'].nunique()}")

📊 Fusion des DataFrames agrégés
🔍 Fusion
   Taille df1 : (2619, 8)
   Taille df2 : (2517, 8)
   Colonnes df1 : ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'Valeur']
   Colonnes df2 : ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'Valeur']
   ✅ Fusion terminée : (2625, 8)
   ✅ Colonnes finales : ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'Valeur']

📋 Aperçu du DataFrame fusionné :


,Code zone,Zone,Code Élément,Élément,Produit,Année,Unité,Valeur
0,1,Arménie,645,Disponibilité alimentaire en quantité (kg/pers...,Total,2013.0,kg,985.68
1,1,Arménie,664,Disponibilité alimentaire (Kcal/personne/jour),Total,2013.0,Kcal/personne/jour,2924.00
2,1,Arménie,674,Disponibilité de protéines en quantité (g/pers...,Total,2013.0,g/personne/jour,90.06
3,1,Arménie,684,Disponibilité de matière grasse en quantité (g...,Total,2013.0,g/personne/jour,96.63
4,1,Arménie,5072,Variation de stock,Total,2013.0,Milliers de tonnes,-95.00
5,1,Arménie,5123,Pertes,Total,2013.0,Milliers de tonnes,228.00
6,1,Arménie,5131,Traitement,Total,2013.0,Milliers de tonnes,166.00
7,1,Arménie,5142,Nourriture,Total,2013.0,Milliers de tonnes,2929.00
8,1,Arménie,5154,Autres utilisations (non alimentaire),Total,2013.0,Milliers de tonnes,242.00
9,1,Arménie,5301,Disponibilité intérieure,Total,2013.0,Milliers de tonnes,4388.00



📊 Statistiques du DataFrame fusionné :
   Lignes : 2625
   Colonnes : 8
   Pays uniques : 175
   Éléments uniques : 15


In [ ]:
## Nettoyage et fusion de fr_population et fr_sousalimentation

def clean_and_merge_nutrition_data(all_dataframes):
    """
    Nettoie et fusionne fr_population et fr_sousalimentation pour créer un DataFrame final
    avec le pourcentage de sous-alimentation.
    
    Args:
        all_dataframes: Dictionnaire contenant tous les DataFrames
    
    Returns:
        DataFrame fusionné avec sous_alim_pct
    """
    import pandas as pd
    
    # Colonnes à conserver
    columns_to_keep = ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'Valeur']
    
    print("🔍 Nettoyage et fusion des données nutritionnelles")
    print("=" * 60)
    
    # Nettoyage de fr_population
    if 'fr_population' in all_dataframes:
        df_pop = all_dataframes['fr_population']
        print(f"   fr_population : {df_pop.shape}")
        
        # Vérification des colonnes
        if not all(col in df_pop.columns for col in columns_to_keep):
            print("   ⚠️ Colonnes manquantes dans fr_population, adaptation nécessaire")
            # On garde les colonnes disponibles
            available_cols = [col for col in columns_to_keep if col in df_pop.columns]
            print(f"   Colonnes conservées : {available_cols}")
        else:
            available_cols = columns_to_keep
            
        # Nettoyage du DataFrame
        df_pop_clean = df_pop[available_cols].copy()
        
        # Suppression des lignes avec valeurs manquantes
        df_pop_clean = df_pop_clean.dropna(subset=['Valeur'])
        print(f"   fr_population nettoyé : {df_pop_clean.shape}")
        
        # Ajout au dictionnaire
        all_dataframes['fr_population_clean'] = df_pop_clean
    
    # Nettoyage de fr_sousalimentation
    if 'fr_sousalimentation' in all_dataframes:
        df_sous = all_dataframes['fr_sousalimentation']
        print(f"   fr_sousalimentation : {df_sous.shape}")
        
        # Vérification des colonnes
        if not all(col in df_sous.columns for col in columns_to_keep):
            print("   ⚠️ Colonnes manquantes dans fr_sousalimentation, adaptation nécessaire")
            # On garde les colonnes disponibles
            available_cols = [col for col in columns_to_keep if col in df_sous.columns]
            print(f"   Colonnes conservées : {available_cols}")
        else:
            available_cols = columns_to_keep
            
        # Nettoyage du DataFrame
        df_sous_clean = df_sous[available_cols].copy()
        
        # Remplacement de l'année '2012-2014' par '2013'
        df_sous_clean['Année'] = df_sous_clean['Année'].replace('2012-2014', '2013')
        
        # Suppression des lignes avec valeurs manquantes
        df_sous_clean = df_sous_clean.dropna(subset=['Valeur'])
        print(f"   fr_sousalimentation nettoyé : {df_sous_clean.shape}")
        
        # Ajout au dictionnaire
        all_dataframes['fr_sousalimentation_clean'] = df_sous_clean
    
    # Fusion des deux DataFrames nettoyés
    if 'fr_population_clean' in all_dataframes and 'fr_sousalimentation_clean' in all_dataframes:
        print("\n🔄 Fusion des DataFrames nettoyés")
        
        df_pop = all_dataframes['fr_population_clean']
        df_sous = all_dataframes['fr_sousalimentation_clean']
        
        # Vérification que les colonnes de jointure existent
        join_cols = ['Code zone', 'Code Élément', 'Année']
        if not all(col in df_pop.columns for col in join_cols):
            print("❌ Colonnes de jointure manquantes dans fr_population_clean")
            return None
        if not all(col in df_sous.columns for col in join_cols):
            print("❌ Colonnes de jointure manquantes dans fr_sousalimentation_clean")
            return None
            
        # Fusion avec outer join pour conserver toutes les données
        merged_df = pd.merge(df_pop, df_sous, on=join_cols, how='outer', suffixes=('_pop', '_sous'))
        
        # Calcul du pourcentage de sous-alimentation
        # (Valeur sous-alim × 1000) / Valeur population × 100
        # Attention : les unités sont différentes (milliers vs millions)
        merged_df['sous_alim_pct'] = (
            (merged_df['Valeur_sous'] * 1000) / 
            (merged_df['Valeur_pop']) * 100
        )
        
        # Correction de l'unité pour la colonne sous_alim_pct
        merged_df['Unité'] = '%'
        
        # Réorganisation des colonnes
        final_cols = ['Code zone', 'Zone', 'Code Élément', 'Élément', 'Produit', 'Année', 'Unité', 'sous_alim_pct']
        existing_final_cols = [col for col in final_cols if col in merged_df.columns]
        merged_df = merged_df[existing_final_cols]
        
        print(f"   ✅ Fusion terminée : {merged_df.shape}")
        print(f"   ✅ Colonnes finales : {list(merged_df.columns)}")
        
        return merged_df
    else:
        print("❌ Impossible de fusionner : DataFrames nettoyés manquants")
        return None

# Exécution du nettoyage et fusion
print("📊 Nettoyage et fusion des données nutritionnelles")
print("=" * 60)

# Création du DataFrame fusionné
final_nutrition_df = clean_and_merge_nutrition_data(all_dataframes)


    
    print("\n📋 Aperçu du DataFrame final :")
    display(final_nutrition_df.head(10))
    
    print(f"\n📊 Statistiques du DataFrame final :")
    print(f"   Lignes : {len(final_nutrition_df)}")
    print(f"   Colonnes : {len(final_nutrition_df.columns)}")
    print(f"   Pays uniques : {final_nutrition_df['Code zone'].nunique()}")
    print(f"   Éléments uniques : {final_nutrition_df['Code Élément'].nunique()}")

📊 Nettoyage et fusion des données nutritionnelles
🔍 Nettoyage et fusion des données nutritionnelles
   fr_population : (175, 9)
   fr_population nettoyé : (175, 8)
   fr_sousalimentation : (1020, 10)
   fr_sousalimentation nettoyé : (605, 8)

🔄 Fusion des DataFrames nettoyés


ValueError: You are trying to merge on int64 and str columns for key 'Année'. If you wish to proceed you should use pd.concat